lua/nvim-treesitter/textobjects/move.lua

```diff
@@ -70,16 +74,157 @@ local function move(query_strings, forward, start, winid)
+      end
+      found = line:find(char, cursor[2] + 2, true)
+    else
+      -- reverse find from the cursor position
+      if not inclusive and repeating then
+        cursor[2] = cursor[2] - 1
+      end
+
+      found = line:reverse():find(char, #line - cursor[2] + 1, true)
+      if found then
+        found = #line - found + 1
+      end
+    end
+
+    if not found then
+      return char
+    end
+
+    if forward then
+      if not inclusive then
+        found = found - 1
+      end
+    else
+      if not inclusive then
+        found = found + 1
+      end
+    end
+
+    cursor[2] = found - 1
+    repeating = true -- after the first iteration, search from the next character if not inclusive.
+  end
+
+  -- move to the found position
+  vim.api.nvim_win_set_cursor(0, { cursor[1], cursor[2] })
+  return char
+end
+
+M.builtin_f = function()
+  local char = builtin_find("forward", "inclusive")
+  if builtin_find ~= nil then
+    M.last_move = { func = builtin_find, args = { "forward", "inclusive", char, "repeating" } }
+  end
+end
+
+M.builtin_F = function()
+  local char = builtin_find(not "forward", "inclusive")
+  if builtin_find ~= nil then
+    M.last_move = { func = builtin_find, args = { not "forward", "inclusive", char, "repeating" } }
+  end
+end
+
+M.builtin_t = function()
+  local char = builtin_find("forward", not "inclusive")
+  if builtin_find ~= nil then
+    M.last_move = { func = builtin_find, args = { "forward", not "inclusive", char, "repeating" } }
+  end
+end
+
+M.builtin_T = function()
+  local char = builtin_find(not "forward", not "inclusive")
+  if builtin_find ~= nil then
+    M.last_move = { func = builtin_find, args = { not "forward", not "inclusive", char, "repeating" } }
+  end
+end
+
+M.repeat_last_move = function()
+  if M.last_move then
+    M.last_move.func(unpack(M.last_move.args))
+  end
+end
+
+M.repeat_last_move_opposite = function()
+  if M.last_move then
+    local args = { unpack(M.last_move.args) } -- copy the table
+    if M.last_move.func == move then
+      args[2] = not args[2] -- reverse the direction
+    else
+      -- builtin_find
+      args[1] = not args[1] -- reverse the direction
+    end
+    M.last_move.func(unpack(args))
+  end
+end
+
+M.repeat_last_move_next = function()
+  if M.last_move then
+    local args = { unpack(M.last_move.args) } -- copy the table
+    if M.last_move.func == move then
+      args[2] = "forward" -- set the direction to forward
+    else
+      -- builtin_find
+      args[1] = "forward" -- set the direction to forward
+    end
+    M.last_move.func(unpack(args))
+  end
+end
+
+M.repeat_last_move_previous = function()
+  if M.last_move then
+    local args = { unpack(M.last_move.args) } -- copy the table
+    if M.last_move.func == move then
+      args[2] = not "forward" -- set the direction to backward
+    else
+      -- builtin_find
+      args[1] = not "forward" -- set the direction to backward
+    end
+    M.last_move.func(unpack(args))
+  end
```

Comment:
Could you explain to me why there are four methods: `repeat_last_move`, `repeat_last_move_opposite`, `repeat_last_move_next` and `repeat_last_move_previous` instead of just having two: `repeat_last_move` and `repeat_last_move_opposite`?

Reply:
`repeat_last_move` goes to the last direction. If you moved backwards, it will move backwards.  
`repeat_last_move_next` ensures that it will move forward.